plugins/out_stackdriver/stackdriver.c

```diff
@@ -331,7 +357,17 @@ static flb_sds_t get_google_token(struct flb_stackdriver *ctx)
          * If the routine fails it will return NULL and the caller will just
          * issue a FLB_RETRY.
          */
-        return oauth2_cache_to_token();
+        output = oauth2_cache_to_token();
+        cached_expiration = oauth2_cache_get_expiration();
+        if (time(NULL) >= cached_expiration) {
+            return output;
+        } else {
+            /* 
+             * Cached token is expired. Wait on lock to use up-to-date token
+             * by either waiting for it to be refreshed or refresh it ourselves.
+             */
+            ret = pthread_mutex_lock(&ctx->token_mutex);
```

Comment:
Changed that to an `if` instead of an `else if` so we don't need to duplicate any logic. Once we acquire the lock, in this case we should do all the rest anyway:
- Check if the token is updated by other threads
- Use it and update the cache if it is updated and exit
- Refresh the token and cache it if not